File: src/dmrgpy/mpsjulialive/excited.py
import numpy as np

from .mpo import MPO


def get_excited_states_dmrg(self,n=2,noise=0.0,scale=10.0):
    """Excited states on the Julia backend, via ITensorMPS.jl's own
    orthogonality-penalty dmrg() (mpsjulialive/excited.jl's
    excited_states_dmrg -- the whole n-state loop runs in one Julia
    call). Mirrors excited.py::get_excited_states_dmrg's C++ path
    (mpscpp2/mpscpp3's Chain::excited_states) and
    pyitensor.chain.Chain.excited_states()."""
    wf0 = self.get_gs()
    H = self.hamiltonian
    e0 = self.e0
    from .dynamics import _max_energy_bound
    emax = _max_energy_bound(self,H) # mutates self.wf0/self.e0, see below
    self.wf0,self.e0,self.computed_gs = wf0,e0,True # restore GS cache
    weight = (emax-e0)*scale
    Hop = MPO(H,MBO=self)
    from .juliasession import Main as Mainjl
    energies_jl,wfs_jl = Mainjl.excited_states_dmrg(Hop.jlmpo,wf0.jlmps,
            int(n),weight,self.jlsites,self.nsweeps,self.cutoff,self.maxm)
    from .mps import MPS
    wfs = [MPS(w,MBO=self) for w in wfs_jl]
    energies = np.array([complex(e).real for e in energies_jl])
    return energies,wfs

File: src/dmrgpy/mpsjulialive/excited.jl

# Excited states via ITensorMPS.jl's own orthogonality-penalty dmrg()
# method (dmrg(H::MPO, Ms::Vector{MPS}, psi0::MPS, sweeps; weight)) --
# mirrors pyitensor/chain.py's excited_states()/dmrg_excited(), which in
# turn mirrors mpscpp3/chain_session.h's Chain::excited_states().

function excited_state_dmrg(H,wfs,psi0,weight,nsweeps,cutoff,maxm;
		noise=0.0)
	# make_sweeps/run_quiet are defined in get_gs.jl, loaded earlier by
	# juliasession.py's initialize() -- shared rather than a third
	# copy-pasted Sweeps-construction+stdout-silencing block.
	sweeps = make_sweeps(nsweeps,maxm,cutoff; noise=noise)
	run_quiet() do
		dmrg(H,wfs,psi0,sweeps;weight=weight)
	end
end


function excited_states_dmrg(H,psi0_gs,n,weight,sites,nsweeps,cutoff,maxm,
		linkdims; noise=0.0)
	# whole n-state loop runs in one Julia call, same design as
	# kpm.jl/tdvp.jl -- one new random-start warm state per additional
	# excited state, deflated against every wavefunction found so far
	# via dmrg()'s own orthogonality-penalty weight. Each start is
	# dmrg_start_state's (mpsalgebra.jl) and the sweeps carry the chain's
	# noise on their first half, as mpscpp3's Chain::excited_states runs
	# them (default_mps() and make_sweeps()); both used to be missing
	# here, each state starting from a product state with no noise
	wfs = MPS[psi0_gs]
	for i=2:n
		psi0 = dmrg_start_state(sites,linkdims)
		e1,psi1 = excited_state_dmrg(H,wfs,psi0,weight,nsweeps,cutoff,maxm;
			noise=noise)
		push!(wfs,psi1)
	end
	# re-evaluate exact <wf|H|wf> for every state (including the ground
	# state) rather than reusing dmrg()'s own penalized eigenvalue
	# estimate -- mirrors mpscpp3/chain_session.h's Chain::excited_states
	energies = ComplexF64[inner(w,H,w) for w in wfs]
	return energies,wfs
end
